Added pgsql cluster option

sean797 opened the pull request:
This adds the ability to setup a HA pgSQL cluster using Streaming Replication.

On a master, you can do:
`--foreman-db-node-type master --foreman-db-cluster-hostnames node2.example.com --foreman-db-host foreman-vip.example.com`

One a slave, you can do:
`--foreman-db-node-type slave --foreman-db-cluster-hostnames node1.example.com --foreman-db-host foreman-vip.example.com --foreman-db-username <SAME AS MASTER> --foreman-db-replication-password <SAME AS MASTER>`

Failover should be just as simple as touching the trigger file. e.g `touch /var/lib/pgsql/data/failover`. You can then set `--foreman-db-node-type master` for consistency.
If the master then recovers you can do `rm -rf /var/lib/pgsql/data/` and then set the machine up as a slave.

By default slaves are asynchronous, slaves can be set to synchronous using the `--foreman-db-synchronous-names` option, though don't do this during initial setup as if the slaves aren't running yet it will result in a error.

ekohl commented on manifests/database/postgresql.pp:
```diff
@@ -10,20 +13,111 @@
+  if $::foreman::db_cluster {
+    $db_all_cluster_hostnames = concat($::foreman::db_cluster_hostnames, $::fqdn)
+    if $::foreman::db_node_type == 'master' {
+      class {'postgresql::server':
```
We try very hard not to manage external classes so users can override them in profiles.

ekohl commented on manifests/database/postgresql.pp:
```diff
@@ -10,20 +13,111 @@
-    encoding => 'utf8',
-    locale   => 'en_US.utf8',
+  } else {
+    class {'::postgresql::client':}
```
We try hard not to manage external classes so users can override them in profiles.

mmoll commented:
I don't think this should be a part of this module, but rather using the puppetlabs-postgresql module to set up the cluster and then conenct it as external DB via puppet-foreman feels the more correct way.

ekohl commented on manifests/database/postgresql.pp:
```diff
@@ -10,20 +13,111 @@
   }
 
+  $rep_password = $::foreman::db_replication_password ? {
+    'UNSET' => false,
```
I strongly dislike the UNSET we already have but dropping them has been too much of an effort until now. Let's at least try to avoid adding more.

ekohl commented on manifests/params.pp:
```diff
@@ -145,6 +157,7 @@
           $plugin_prefix = 'tfm-rubygem-foreman_'
           $init_config = '/etc/sysconfig/foreman'
           $init_config_tmpl = 'foreman.sysconfig'
+          $postgres_home = "/var/lib/pgsql${::postgresql::params::package_version}"
```
Won't this fail if the params aren't included?

ekohl commented on manifests/database/postgresql.pp:
```diff
@@ -10,20 +13,111 @@
+    } elsif $::foreman::db_node_type == 'slave' {
+      file { 'Write .pgpass':
+        content => "*:*:*:${::foreman::db_replication_username}:${::foreman::db_replication_password}\n",
+        path    => "${::foreman::params::postgres_home}/.pgpass",
```
It surprises me a little that there's no variable for this in puppetlabs-postgresql.

sean797 replied:
Yes, I was also - really annoying. I'll send a PR to them this week.

ekohl commented on manifests/database/postgresql.pp:
```diff
@@ -10,20 +13,111 @@
+    }
   }
 
-  Postgresql::Server::Role[$::foreman::db_username] -> Postgresql::Server::Database[$dbname]
```
This chaining was added in puppetlabs-postgresql 4.1.0. Please also increase the minimum version in metadata.json.

ekohl replied:
Created https://github.com/theforeman/puppet-foreman/pull/546 to address this. It was also 4.2.0.

ekohl commented on manifests/init.pp:
```diff
@@ -159,6 +159,24 @@
+#                               type:Optional[Array]
+#
+# $db_synchronous_names::       List of hostnames in the cluster using synchronous replication.
+#                               type:Optional[Array]
```
I think this could even be ```Optional[Array[String]]```

ekohl commented on manifests/init.pp:
```diff
@@ -159,6 +159,24 @@
+#                               type:Optional[String]
+#
+# $db_cluster_hostnames::       List of hostnames in the cluster. Excluding this one.
+#                               type:Optional[Array]
```
I think this could even be ```Optional[Array[String]]```

ekohl commented on manifests/init.pp:
```diff
@@ -159,6 +159,24 @@
+# $db_cluster::                 Setup a database cluster. PostgreSQL only.
+#                               type:Boolean
+#
+# $db_node_type::               Cluster node type, can be one of 'master' or 'slave'.
```
Would it make sense to have a an explicit name for the standalone version?

ekohl commented on manifests/database/postgresql.pp:
```diff
@@ -10,20 +13,111 @@
-  include ::postgresql::client, ::postgresql::server
+  if $::foreman::db_cluster {
+    $db_all_cluster_hostnames = concat($::foreman::db_cluster_hostnames, $::fqdn)
+    if $::foreman::db_node_type == 'master' {
```
Does it make sense to split this into ```foreman::database::postgresql::master``` and ```foreman::database::postgresql::slave```? Maybe just ```foreman::database::postgresql::cluster```?

sean797 commented:
@mmoll I totally agree the better way to scale is using an external DB, but I think is defiantly valuable to be able to deploy a full self-maintained foreman cluster. Not everyone wants to build/manage/maintain an external DB.

@ekohl I think I addressed all your comments, also added some tests. Thanks for the quick review 👍

ekohl commented on manifests/database/postgresql/master.pp:
```diff
@@ -0,0 +1,46 @@
+# Set up the foreman master database using postgresql
+class foreman::database::postgresql::master {
+  $db_all_cluster_hostnames = concat($::foreman::db_cluster_hostnames, $::fqdn)
```
Maybe a ```unique()``` around it as well to be sure?

ekohl commented on manifests/database/postgresql/master.pp:
```diff
@@ -0,0 +1,46 @@
+    locale   => 'en_US.utf8',
+  }
+
+  ::foreman::database::postgresql::db_access { $db_all_cluster_hostnames: }
```
Defines don't need the ```::``` prefix.

ekohl commented on manifests/database/postgresql/slave.pp:
```diff
@@ -0,0 +1,58 @@
+  }
+
+  postgresql::server::config_entry { 'wal_keep_segments':
+    value => '256',
```
Would it make sense to make this a class parameter (with the same default)?

ekohl commented on manifests/database/postgresql/slave.pp:
```diff
@@ -0,0 +1,58 @@
+class foreman::database::postgresql::slave {
+  require ::postgresql::params
+
+  $db_all_cluster_hostnames = concat($::foreman::db_cluster_hostnames, $::fqdn)
```
Use a ```unique()``` as well?

sean797 commented:
Humm a `puppet-foreman_profiles` module could be useful in the future..

Anyway I've updated this and tests passed now 👍
